Lib/typing.py

```diff
@@ -280,25 +284,51 @@ def _collect_parameters(args):
                         parameters.append(collected)
         elif hasattr(t, '__typing_subst__'):
             if t not in parameters:
+                if type_var_tuple_encountered and t.__default__ is not None:
+                    raise TypeError('Type parameter with a default'
+                                    ' follows TypeVarTuple')
+
+                if t.__default__ is not None:
+                    default_encountered = True
+                elif default_encountered:
+                    raise TypeError(f'Type parameter {t!r} without a default'
+                                    ' follows type parameter with a default')
+
                 parameters.append(t)
         else:
+            if _is_unpacked_typevartuple(t):
+                type_var_tuple_encountered = True
             for x in getattr(t, '__parameters__', ()):
                 if x not in parameters:
                     parameters.append(x)
     return tuple(parameters)
 
 
-def _check_generic(cls, parameters, elen):
+def _check_generic_specialization(cls, parameters):
     """Check correct count for parameters of a generic cls (internal helper).
 
     This gives a nice error message in case of count mismatch.
     """
+    elen = len(cls.__parameters__)
     if not elen:
         raise TypeError(f"{cls} is not a generic class")
     alen = len(parameters)
     if alen != elen:
-        raise TypeError(f"Too {'many' if alen > elen else 'few'} arguments for {cls};"
-                        f" actual {alen}, expected {elen}")
+        # deal with defaults
+        if alen < elen:
+            # since we validate defaults in _collect_type_vars
+            # or _collect_parameters, we can safely check parameters[alen]
+            if cls.__parameters__[alen].__default__ is not None:
+                return
+
+            elen -= sum(p.__default__ is not None for p in cls.__parameters__)
+            expect_val = f"at least {elen}"
+        else:
+            expect_val = elen
+
+        raise TypeError(f"Too {'many' if alen > elen else 'few'} arguments"
+                        f" for {cls}; actual {alen}, expected {expect_val}")
```

Comment:
I'd find this function so much more readable if we renamed `alen` and `elen` to `actual_len` and `expected_len`, respectively. It felt like it took me way too long to refresh my memory about what this function did :-(

Reply:
Either that or `len/maxlen` (because, when you have default parameters, this doesn't mean that you need to specify everything).

Reply:
I'm going with Alex's suggestion and I'll also rename "parameters" to "arguments" (parameters are the things in the definition, arguments are what you pass when you do a specialization).